### DISCUSSION TITLE
Wie viele Pylontech Batterien kann ich anschließen?

### DISCUSSION THREAD

**Discussion by Manos1966** [skippermeister/OpenDTU-OnBattery#16] [Q&A]
Hi @skippermeister 

Wieviele angeschlossene Pylontech Batterien kann ich per RS485 lesen?

Wie waere es mit 20? 🙈 

... und wird es helfen wenn ich einen ESP32-S3 N16R8 anstatt einen ESP32 4MB nutze?

**Comment by skippermeister** [comment]
Hallo @Manos1966,

der initiale Scan in fragt maximal den Master plus 7 Slave Batterien ab. Der Scan zeigt im seriellen Log die Anzahl der gefundenen Batterien, jedoch maximal 8 Batterien. Falls mehr als 8 Batterien angeschlossen sind, werden diese nicht mehr im initialen Scan erkannt.

Zugelassen habe ich allerdings nur die Master Batterie und eine Slave Batterie. Falls mehr als eine Slave Batterie  abgeschlossen sind, werden diese ignoriert.
Ich habe keine Möglichkeit zu testen, was bei mehr als in Summe 2 Batterien passiert. Das ganze wird vom benötigten RAM-Speicher beschränkt.

Die Batterien werden nacheinander abgefragt. Pro Scheduler Durchlauf etwa eine Batterie. Somit verzögert sich die gesamte Abfrage aller angeschlossenen Batterien entsprechend z.B. bei 20 Batterien um den Faktor 10. Ich sehe das allerdings nicht als ein Problem.

In wieweit die RAM-Resourcen aus dem PSRAM genutzt werden, kann ich nicht sagen. Hast Du mehr als zwei Batterien ?
Ich kann ohne weiteres eine Image mit 16 oder 20 Batterien erstellen.
Wer testet es dann ?

**Reply by Manos1966** [reply]
Als erstes: VIELEN DANK Andreas, dass du immer eine so positive Denkweise und Hilfsbereitschaft hast 💯  

> Ich kann ohne weiteres eine Image mit 16 oder 20 Batterien erstellen.

❤️ JA BITTE! Fuer eine 15 Batterien Version wäre ich Dankbar!
.....🤔 🤔 🤔 🤔 🤔 Wie viele Kasten Bier wird es mir kosten? 😆 

Situation: 
Ein Bekannter hat 15 x US3000 zusammen und 5 x US2000 zusammen. Und er hat "Anfangsverdacht"...

Ich habe im Betrieb eine 24kW Anlage mit 10 x US3000 wo ich testen kann. 
Ich könnte sogar den Unterschied zwischen ESP32 und ESP32-S3 N16R8 testen wenn du willst.
![20241011_180314 - small](https://github.com/user-attachments/assets/a4cbbef6-9c8e-4255-b005-362eaf5f4ab5)

**Reply by skippermeister** [reply]
Hallo @Manos1966,

ich habe mich mal mit dem RS485 Protokoll der Pylontech Batterie auseinandergesetzt. Die Batterien werden in Gruppen über RS485 zusammengeschaltet. Maximal können 8 Gruppen mit je bis zu 11 Slave Batterien gekoppelt werden.

Die Gruppen Adressen werden über die DIP Schalter der jeweiligen Master Batterie einer Gruppe definiert.
Die RS485 Adressen der Batterien bilden sich dann wie folgt (so habe ich es verstanden):

Master Gruppen Adresse = GruppenNummer  x  0x10 + 2 (0x02 für z.B. den Master in Gruppe 0 , 0x32 für Master in Gruppe 3)
Slave Adresse = GruppenNummer  x  0x10 + SlaveNummer+2 (0x47 für z.B. den 5. Slave in Grupp 4)

Jetzt kommt allerdings mein Problem. Meine Routine zur Abfrage und Identifizierung der Batterien kennt nur die Gruppe 0 (Single Group Mode). Die Routine zur Identifizierung (Init Phase) fängt in der Gruppe 0 von Master Adresse 2 an und geht hinauf bis Adresse 9 (=7 Slave Batterien). Das kann ich erweitern auf 11 Slaves.
Die Abfrage der Batterien und Anzeige ist aktuell auf ein Master und ein Slave begrenzt.
Ich würde in der ersten Version auf ein Master (Gruppe 0, Single Group) und 11 Slaves erweitern.

Der schwierigere Teil ist die Änderung in der Web Gui. Die gesamte Anzeige in der Web Gui ist auf einen einzigen Master mit mehreren Slaves ausgelegt. Ich weiß noch nicht wie ich das umsetze kann.

**Reply by Manos1966** [reply]
Hi @skippermeister 

> Ich würde in der ersten Version auf ein Master (Gruppe 0, Single Group) und 11 Slaves erweitern.

Das reicht vollkommen aus! 
Wie viele Leute werden mehr als 12 Batterien (Master + 11 Slaves) haben?
Ich habe zehn auf die Notstrom-Anlage und die Verkabelung der "schwachen" US3000 ist ein Monster! (Das System wurde fertig montiert im Schrank geliefert)

zum Thema Gruppen, wenn jemand mehrere Gruppen hat, der kann sich auch die Kosten eines extra ESP32 mit RS485 Modul leisten....

> Die gesamte Anzeige in der Web Gui ist auf einen einzigen Master mit mehreren Slaves ausgelegt.

Das reicht vollkommen aus! 

Die Frage die ich jetzt habe: Sollte ich einen ESP32 testen, oder sollte ich direkt einen ESP32-S3 probieren? (wegen des geringeren Hauptspeichers des ESP32)
Bin am loetten... 🙈

**Reply by skippermeister** [reply]
@Manos1966 ,
Ich bin gerade am kompilieren und testen.
Ich würde gleich auf ESP32-S3 mit 16N8 gehen. Welches Modell hast Du? Es gibt Unterschiede in Bezug auf die PSRAM SPI Ansteuerung 1.8Volt oder 3.3Volt. Mein platformio.ini Template berücksichtigt aktuell die 3.3 Volt Variante.

**Reply by Manos1966** [reply]
ESP32-S3 N16R8
It should be 3,3V

If I remember correctly, only the ones with the letter V at the end are 1,8V
 
![image](https://github.com/user-attachments/assets/4a274ea2-1a79-49d6-b04b-8eee652dc8d8)

**Reply by Manos1966** [reply]
Hallo Andreas @skippermeister 
>Ich bin gerade am kompilieren und testen

kommst du voran? Ich habe beide Versionen (ESP32/ESP32-S3) fertig zum testen...

**Reply by Manos1966** [reply]
Hmmm... ich brauche ein bisschen Hilfe 🙈 

Letztes offizielles Release 2024.10.25
ESP32-S3 N16R8
Firmware Version generic_esp32s3_meanwell
![image](https://github.com/user-attachments/assets/6869d668-ddfe-4d06-ad35-e44ff114b86b)

Zwei Pylontechs definiert.
Ich sehe die MASTER (siehe Bild), aber keine Spalte Links wo ich zwischen Master/Slave wechseln koennte.
![image](https://github.com/user-attachments/assets/30293fc2-c943-4179-85dc-3e9678784601)

**Reply by Manos1966** [reply]
Just to be sure I am not dreaming, my old ESP32 from Oktober (?) 2023 is working correctly. I see Master and Slave:
![image](https://github.com/user-attachments/assets/b0dcb5aa-8f4a-408a-a712-f9fd068a1a3d)

![image](https://github.com/user-attachments/assets/f771c387-00e7-493f-afb7-0cf60c354535)

**Reply by skippermeister** [reply]
Hi @Manos1966,

den Fehler, warum nur eine Batterie angezeigt wird habe ich gefunden. 

Die Software ist zwischenzeitlich erheblich erweitert worden und hat auch einen Großteil der aktuellen Updates von der original openDTU-onBattery drin.

Ich will noch nicht die neue Version mit der Erweiterung auf 12 Batterien im Source Code im github committen. Ich habe fertige factory firmware.bin und firmware.bin fuer ESP32-S3 16N8 fertig. Ich würde diese jetzt Dir zum Test geben wollen.

**Reply by Manos1966** [reply]
... ich glaube auch nicht, dass jemals jemand so viele Batterien anschliessen wird 🙈

**Reply by Manos1966** [reply]
Soll ich dir meine Email Adresse geben?

Andere Baustelle, aber bevor ich es vergesse, dieses Projekt wirst du Interessant finden (fuer deine zukünftigen HUAWEI Forschungen)
https://github.com/mjpalmowski/CAN-BUS-control-R4875G1-with-ESPHome-and-MQTT
Steuerung fuer HUAWEI R4875G, R4850G, R4830G

**Reply by skippermeister** [reply]
@Manos1966 
hier findest Du die bin Dateien. Ich habe es doch ins github hochgeladen
https://github.com/skippermeister/OpenDTU-OnBattery/actions/runs/12433582011

**Reply by skippermeister** [reply]
@Manos1966,
bitte ein log erstellen, und zwar logs für Batterie aktivieren. Die anderen bitte deaktivieren, damit ich den Überblick behalte.
Ich benötige die logs vom Start des ESP32 bis ca. 2min nach dem Start. In dieser Zeit scanned das Pylontech RS485 Modul den RS485 Bus auf angeschlossene Batterien (Gruppenweise und jede Gruppe auf verkettete Batterien).
Danach werden 12 Batterien (Master der Gruppe 0 + 11 Salves der Gruppe 0) im normalen Betrieb versucht auszulesen.

Der interessante Bereich für mich sieht in etwa so aus (bei mir findet er keine Batterie, weil ich auch an meinem Test ESP32-S3 16N8 keine angeschlossen habe)

[Pylontech RS485]::readParameter: Read basic infos and parameters.
16:11:15.090 > [Pylontech RS485]::get_pack_count REQUEST & GET Group 0 Module 2 InfoCommand 02
16:11:15.090 > [Pylontech RS485]:send_cmd frame=~20024690E00202FD30
16:11:17.176 > [Pylontech RS485] RS485 read timeout
16:11:17.176 > [Pylontech RS485] Battery not responding, check cabling or battery power switch
16:11:17.675 > [Pylontech RS485]::get_pack_count REQUEST & GET Group 1 Module 2 InfoCommand 12
16:11:17.675 > [Pylontech RS485]:send_cmd frame=~20124690E00212FD2E
16:11:19.775 > [Pylontech RS485] RS485 read timeout
16:11:19.775 > [Pylontech RS485] Battery not responding, check cabling or battery power switch
16:11:20.278 > [Pylontech RS485]::get_pack_count REQUEST & GET Group 2 Module 2 InfoCommand 22
16:11:20.278 > [Pylontech RS485]:send_cmd frame=~20224690E00222FD2C
16:11:22.382 > [Pylontech RS485] RS485 read timeout
16:11:22.382 > [Pylontech RS485] Battery not responding, check cabling or battery power switch
16:11:22.884 > [Pylontech RS485]::get_pack_count REQUEST & GET Group 3 Module 2 InfoCommand 32
16:11:22.884 > [Pylontech RS485]:send_cmd frame=~20324690E00232FD2A
16:11:24.977 > [Pylontech RS485] RS485 read timeout
16:11:24.977 > [Pylontech RS485] Battery not responding, check cabling or battery power switch
16:11:25.473 > [Pylontech RS485]::get_pack_count REQUEST & GET Group 4 Module 2 InfoCommand 42
16:11:25.489 > [Pylontech RS485]:send_cmd frame=~20424690E00242FD28
16:11:27.578 > [Pylontech RS485] RS485 read timeout
16:11:27.578 > [Pylontech RS485] Battery not responding, check cabling or battery power switch
16:11:28.073 > [Pylontech RS485]::get_pack_count REQUEST & GET Group 5 Module 2 InfoCommand 52
16:11:28.089 > [Pylontech RS485]:send_cmd frame=~20524690E00252FD26
16:11:30.175 > [Pylontech RS485] RS485 read timeout
16:11:30.188 > [Pylontech RS485] Battery not responding, check cabling or battery power switch
16:11:30.688 > [Pylontech RS485]::get_pack_count REQUEST & GET Group 6 Module 2 InfoCommand 62
16:11:30.688 > [Pylontech RS485]:send_cmd frame=~20624690E00262FD24
16:11:32.788 > [Pylontech RS485] RS485 read timeout
16:11:32.788 > [Pylontech RS485] Battery not responding, check cabling or battery power switch
16:11:33.288 > [Pylontech RS485]::get_pack_count REQUEST & GET Group 7 Module 2 InfoCommand 72
16:11:33.288 > [Pylontech RS485]:send_cmd frame=~20724690E00272FD22
16:11:35.388 > [Pylontech RS485] RS485 read timeout
16:11:35.388 > [Pylontech RS485] Battery not responding, check cabling or battery power switch
16:11:35.888 > [Pylontech RS485] Found 0 groups with in total 0 Battery Packs
16:11:35.889 > [Pylontech RS485]::readParameter done
16:11:35.889 > [Pylontech RS485] Master Battery Pack 2
16:11:35.889 > [Pylontech RS485] request time 0 ms, Cmnd: 42

Ich bin gespannt ob es funktioniert.

**Reply by Manos1966** [reply]
Erster Test, auf zwei Pylon US5000:
PIO Environment	generic_esp32s3_16mb_psram_usb_huawei
[First Log - two Pylons.txt](https://github.com/user-attachments/files/18212814/First.Log.-.two.Pylons.txt)
Test erfolgreich, sogar besser als es mir Lieb war 🤮 
![Problem2](https://github.com/user-attachments/assets/b7eb83f4-8f26-4696-9b10-bfde9a969ae0)


@skippermeister  
Zweiter Test mit zehn US3000: NICHT ERFOLGREICH
vielleicht weil ich die DTU nicht im WiFi gemeldet hatte und dadurch kein richtiges Datum/Zeit ?? 🤔 
[Test2.txt](https://github.com/user-attachments/files/18214562/Test2.txt)

I did not have too much time, I will try again tomorrow (with WiFi).
![image](https://github.com/user-attachments/assets/f5c628f6-a504-49f8-a5b8-bcfabedbc7b7)
![image](https://github.com/user-attachments/assets/c5eb6e29-cf4c-41cd-8024-c264f09f755b)

**Reply by skippermeister** [reply]
@Manos1966,
Aus dem Log geht hervor, dass beim scannen die Master Batterie sich nicht meldet. Es könnte ein Timeout sein, weil die Master Batterie zu lange benötigt um die Slaves abzufragen. Zwei Batterien funktioniert ja.
Versuche deshalb zunächst mit 3, dann in weiteren Schritten mit 5, 7, 9 Batterien zu testen.

**Reply by skippermeister** [reply]
Wie hast Du die DIP Schalter der Master Batterie und der Slave Batterien eingestellt ?

**Reply by Manos1966** [reply]
ich werde spaeter testen...

**Comment by Manos1966** [comment]
Hallo Andreas,

Frohes neues Jahr!

Die erste Gruppe mit 5 x US3000 funktioniert einwandfrei (Bilder folgen).

Nach den Ferien werden die 15 x US2000 getestet 🙈 🙈🙈🙈

Es ärgert mich, dass mein Schrank mit 10 x US3000 nicht mit RS485 funktioniert. Ich habe erfahren, dass diese 10 x US3000 speziell konfiguriert sind vom SystemHaus (es ist ein Schlüsselfertiges Notstrom System), daher darf ich nichts anfassen 🤮

**Reply by skippermeister** [reply]
@Manos1966,

ich wünsche auch Dir ein Frohes Neues Jahr.

Das klingt schon ganz zuversichtlich, aber beachte:
Es sollten mehr als 12 Batterien beim SCAN erkannt werden, jedoch
in einer Gruppe können maximal 12 Batterien enthalten sein (1x Master + 11 Slave, das ist von Pylontech so vorgesehen).
Die Gruppen werden über den DIP Switch der Master Batterie eingestellt. 
Ich meine die Master Batterien müssen dann über RS485 verbunden werden.

Die eigentliche Abfrage der Batterien erfolgt nach dem SCAN allerdings nur auf die Master Batterie der Gruppe 0 und die daran anhängenden maximal 11 Slave Batterien.

Wenn Du mehrere Gruppen bildest, dann schicke mir dazu das Log der ersten 5 Minuten nach dem Reboot des ESP32. Ich kann daraus erkennen, ob der SCAN über die Gruppen funktioniert und aus dem Ergebnis einen Algorithmus erstellen um über mehrere Gruppen die Batterien abzufragen.

Zu Deiner  Notstrom Batterie mit 10x US3000C, wie sind die DIP Switches der Batterien eingestellt? Aus Deinen Bildern kann ich entnehmen, dass an der obersten Batterie am Port A/CAN oder B/RS485 ein RJ45 Kabel steckt und die Batterien über Linkport0 und 1 per Daisychain verbunden sind. Was steckt an dem A/CAN oder B/RS485 Port ? Ist da eine OpenDTU dran ?

**Reply by Manos1966** [reply]
> Zu Deiner Notstrom Batterie mit 10x US3000C, wie sind die DIP Switches der Batterien eingestellt? 

Alle gleich, alle OFF
![image](https://github.com/user-attachments/assets/caad8d59-482d-4b9f-ba11-c5a857080b1b)

> Aus Deinen Bildern kann ich entnehmen, dass an der obersten Batterie am Port A/CAN oder B/RS485 ein RJ45 Kabel steckt und die Batterien über Linkport0 und 1 per Daisychain verbunden sind. 

An dem A/CAN Port steckt die Verbindung zum Victron GX
An dem B/RS485 Port ist die OpenDTU mit der RS485 Software
über Linkport0 und 1 per Daisychain verbunden sind die Batterien

Die Batterien muessten sogar in bestimmter Rheinfolge im Rack platziert werden! Als wir sie verkehrt montiert haben, hat das System nicht funktioniert 🙈 

Bezueglich Batterie Grouppen, ich weiss nicht ob du das hast:
![image](https://github.com/user-attachments/assets/9ec6fedb-85c7-4d5e-914e-9abd0f1b2536)

**Reply by skippermeister** [reply]
Ja, das kenne ich aus dem Handbuch zur Batterie. Widersprüchlich dazu ist die Beschreibung in der RS485 Protokollbeschreibung
![image](https://github.com/user-attachments/assets/87ab7948-c658-4b48-aea5-060de81ccd86)

und hiernach sollen es 8 Gruppen zu je 12 Batterien (1x Master je Group und 12 Slaves je Group)  sein.
![image](https://github.com/user-attachments/assets/c608d655-031c-441e-a2ce-7b29c3e3917f)
 
Da können wir jetzt nur probieren.

**Comment by Manos1966** [comment]
Hallo @skippermeister 
RESPEKT!
Anbei der erste Test. 12 Batterien (US2000)
Funktioniert wie erwartet 🥂 

![00 Working proof](https://github.com/user-attachments/assets/5dd619e5-01ad-451b-bc11-d25673bf3e1d)

Es sind insgesamt 15 angeschlossen.
Sollten wir auch die restlichen 3 probieren?
Wenn ich deine Bilder korrekt verstanden habe, 
- wir muessen die Master Batterie DIP 2 aktivieren (= First Group)
- und die dreizehnte DIP3 aktivieren (=Second Group)

Richtig? 

![image](https://github.com/user-attachments/assets/235209a6-2916-4c7a-a368-effd43f244ed)

**Reply by Manos1966** [reply]
PS. und ich brauche eine Adresse um den Bierkasten zu liefern....

Du kannst mir eine private Nachricht in akkudoktor.net schicken 
(fast gleicher Usernamen, Manos66 )

**Comment by Manos1966** [comment]
@skippermeister Das mit dem Bierkasten war ernst gemeint..... 🙈
